SpellChecker.java

```diff
@@ -11,25 +11,65 @@ public static void main(String[] args) {
 	}
 
 	public static String tail(String str) {
-		// Your code goes here
+		return str.substring(1);
 	}
 
 	public static int levenshtein(String word1, String word2) {
-		// Your code goes here
+		word1 = word1.toLowerCase();
+        word2 = word2.toLowerCase();
+
+		int word1Length = word1.length();
+		int word2Length = word2.length();
+
+		if (word1Length == 0 && word2Length == 0) {
+			return 0;
+		}
+		if (word1Length == 0) {
+			return word2Length;
+		}
+		if (word2Length == 0) {
+			return word1Length;
+		}
+		if (word1.charAt(0)== word2.charAt(0)) {
+			return levenshtein(tail(word1), tail(word2));
+
+		}
+		int insertCost = levenshtein(tail(word1), word2) + 1;
+        int deleteCost = levenshtein(word1, tail(word2)) + 1;
+        int replaceCost = levenshtein(tail(word1), tail(word2)) + 1;
+
+		return Math.min(insertCost, Math.min(deleteCost, replaceCost));
 	}
 
 	public static String[] readDictionary(String fileName) {
 		String[] dictionary = new String[3000];
 
 		In in = new In(fileName);
-
-		// Your code here
+		for (int i = 0 ; i < dictionary.length ; i++) {
+			dictionary[i] = in.readString();
+		}
 
 		return dictionary;
 	}
 
 	public static String spellChecker(String word, int threshold, String[] dictionary) {
-		// Your code goes here
+		int tempDistance;
+		int minDistance = word.length();
+		String resemblesWord = "";
+	
+		for (int i = 0; i < dictionary.length; i++) {
+			tempDistance = levenshtein(word.toLowerCase(),dictionary[i]);
+			if (tempDistance < minDistance) {
+			    minDistance = tempDistance;
+			    resemblesWord = dictionary[i];
+			}
+		}	
+			if (minDistance <= threshold) {
+				return resemblesWord;
+			} else {
+				return word;
+			}
```

Comment:
When your `if` returns value, `else` statement is not required since the method would exist before it arrive to that line of code.
```suggestion
			return word;
```